### PR TITLE
Timeout Injection Task

### PR DESCRIPTION
**Pull request by esnible**
This describes how to specify a timeout.
I did it as a complete end-to-end, showing a user how to set up a frontend and slow backend, and then how to add an Istio rule to introduce a timeout.
Like the Bookinfo Sample, this uses .yaml files to describe the Services and Deployments.  Currently thi means we need to clone the images.  Perhaps it would be better to follow the format in https://kubernetes.io/docs/tutorials/connecting-apps/connecting-frontend-backend/ and give the URL so that the user doesn't need to clone anything to practice doing the task.

### CODE REVIEW

File: docs/tasks/timeout-injection.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
+# Injecting Timeouts
```
Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: This document needs to be updated to fit in istio.io's model. You can read about what's needed here: https://istio.io/docs/home/contribute/writing-a-new-topic.html

Basically, the doc needs to include Jekyll front-matter, needs to use the templating mechanism, and the document needs to be registered in the navigation side-bar file.

Thanks.

File: docs/tasks/timeout-injection.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+![Timeout Injection](timeout-injection.png)
+
+Slow or unresponsive microservices can reveal defects in other system components.  Unhandled timeout
+exceptions can cascade and affect user experience.
+
+In this tutorial we will show advanced features of the Istio command line to inject timeouts into
```
Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: This task shows advanced ...

File: docs/tasks/timeout-injection.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+a running system.
+
+The Kubernetes-based implementation of Istio used in this tutorial works by injecting behavior
+into Pods.  All Pods that participate in the service mesh are enabled by the addition of a proxy.
```
Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: Pods -> pods

File: docs/tasks/timeout-injection.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+The _istioctl kube-inject_ subcommand can be used to apply the changes to a running Deployment.
```
Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: Deployment -> deployment

File: docs/tasks/timeout-injection.md (hunk 2 above)
Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: Add a reference to the istioctl command-line reference page.

File: docs/tasks/timeout-injection.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+Later we will show how to modify the YAML files used to deploy your application so that it comes
+up immediately with the Istio mesh connection.
+
+## Before you begin
+
+This tutorial assumes you have deployed Istio on Kubernetes.  If you have not done so, please first
+follow [the installation instructions](../INSTALL.md).
```
Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: Update link.

File: docs/tasks/timeout-injection.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+The Kubernetes-based implementation of Istio used in this tutorial works by injecting behavior
+into Pods.  All Pods that participate in the service mesh are enabled by the addition of a proxy.
+The _istioctl kube-inject_ subcommand can be used to apply the changes to a running Deployment.
+Later we will show how to modify the YAML files used to deploy your application so that it comes
+up immediately with the Istio mesh connection.
+
+## Before you begin
+
+This tutorial assumes you have deployed Istio on Kubernetes.  If you have not done so, please first
```
Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: This task

File: docs/tasks/timeout-injection.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+follow [the installation instructions](../INSTALL.md).
+
+## Connecting microservices with Istio
+
+This guide shows how to set up Istio and manipulate the proxy mesh to achieve useful behavior.
```
Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: This task

Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: proxy mesh -> service mesh

File: docs/tasks/timeout-injection.md [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+In this example we use *[httpbin](https://httpbin.org)* and *NGINX* as and demonstrate how to manipulate them:
```
Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: Bad fragment: "as and"

File: docs/tasks/timeout-injection.md [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+* [citizenstig/httpbin](https://hub.docker.com/r/citizenstig/httpbin/) is a popular Docker image for httpbin
+* NGINX is the offical NGINX Docker image.  We will configure it to redirect all calls to httpbin
```
Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: Missing .

File: docs/tasks/timeout-injection.md [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+First, let's start httpbin and have it join the Istio service mesh
+
+```
+# Start the citizenstig/httpbin image the container in a Kubernetes Pod.  8000 is the httpbin port.
```
Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: Pod -> pod

Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: Bad fragment: "image the container"

File: docs/tasks/timeout-injection.md [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+# Give the file to istioctl
+istioctl create -f /tmp/httpbin-3s-rule.yaml
+```
+
+Wait a few seconds, then issue `time curl -i $IP:$NGINX_NODEPORT/delay/5` again.  This time the response appears after
+three seconds.  Although _httpbin_ was waiting 5 seconds, Istio cut off the request at 3 seconds.
+
+## Cleanup
+
+To remove the rules, deployments and services used in this tutorial:
```
Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: tutorial -> task

File: docs/tasks/timeout-injection.md [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+```
+istioctl delete route-rule httpbin-3s-rule
+kubectl delete -f httpbin.yaml
+kubectl delete -f nginx-httpbin.yaml
+```
+
+
+## Discussion
+
+We have seen two web services, httpbin and NGINX, connected via an Istio proxy mesh.  We have
```
Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: web services -> microservices

File: docs/tasks/timeout-injection.md [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+used the proxy mesh to introduce a maximum delay using a timeout in all communications to httpbin.
```
Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: proxy mesh -> service mesh.

File: docs/tasks/timeout-injection.md [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+
+## Next Steps
```
Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: Should we using the template for this, will come out as What next instead.

File: docs/tasks/timeout-injection.md [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+
+For a more detailed example, showing version upgrades and custom routing based on HTTP headers, try the
+[bookinfo demo](../demos/apps/bookinfo/README.md).
```
Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: Fix link.

File: docs/tasks/timeout-injection.md (hunk 5 above)
Comment (esnible) [inline docs/tasks/timeout-injection.md]: Frank's _bookinfo.md_ shows how to _git_ clone _istio/istio_ and bring up Istio.  Are there instructions that I am missing or should every page show how to bring up Istio?

File: docs/tasks/timeout-injection.md (hunk 7 above)
Comment (esnible) [inline docs/tasks/timeout-injection.md]: Do we have a glossary page?  A list of all the terms would help both users and doc authors.

File: docs/tasks/timeout-injection.md (hunk 14 above)
Comment (esnible) [inline docs/tasks/timeout-injection.md]: I am not sure how to use the template, so I removed the Next Steps section.

**Comment by esnible** [comment]
I have made the requested changes.

Is there any way that I can build the .html files from the .md so that I can test what I have written before I do future documentation pushes?

File: docs/tasks/timeout-injection.md (hunk 5 above)
Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: Pulling that out into a separate doc and referencing it from both places would be a good idea.

File: docs/tasks/timeout-injection.md (hunk 7 above)
Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: Excellent idea. I'll figure out where to put such a thing and seed it with a few things to get the ball rolling.

File: docs/tasks/timeout-injection.md (hunk 14 above)
Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: See here for instructions on how to use the templates:

https://istio.io/docs/home/contribute/using-page-templates.html

File: docs/tasks/timeout-injection.md [hunk 16]
```diff
@@ -0,0 +1,120 @@
+---
+title: Timeout Injection
+headline: 'Timeout Injection walkthrough'
+sidenav: doc-side-tasks-nav.html
+bodyclass: docs
+layout: docs
+type: markdown
+---
+# Injecting Timeouts
```
Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: Please remove this line, we already inject titles on the page based on the Jekyll front matter.

File: docs/tasks/timeout-injection.md [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+The Kubernetes-based implementation of Istio used in this task works by injecting behavior
+into pods.  All pods that participate in the service mesh are enabled by the addition of a proxy.
+We will inject HTTP timeout behavior by giving a new rule to the Istio Manager.  The proxy will
+implement the timeout.
+
+## Before you begin
+
+This task assumes you have deployed Istio on Kubernetes.  If you have not done so, please first
+follow [the installation instructions](../TBD.md).  You should also have installed
+the [istioctl](../reference/istioctl.md) CLI
```
Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: Use this instead:
```
[istioctl]({{site.baseurl}}/reference/istioctl.html)
```

File: docs/tasks/timeout-injection.md [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+
+The Kubernetes-based implementation of Istio used in this task works by injecting behavior
+into pods.  All pods that participate in the service mesh are enabled by the addition of a proxy.
+We will inject HTTP timeout behavior by giving a new rule to the Istio Manager.  The proxy will
+implement the timeout.
+
+## Before you begin
+
+This task assumes you have deployed Istio on Kubernetes.  If you have not done so, please first
+follow [the installation instructions](../TBD.md).  You should also have installed
```
Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: Is this TBD intentional?

File: docs/tasks/timeout-injection.md (hunk 1 above)
Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: Still need to use the templates. If the {% capture %} {% endcapture %} annotations you can see in the other documents in istio.github.io.

File: docs/tasks/timeout-injection.md [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+Next we will start NGINX and configure it to proxy for the httpbin service started in the previous step
+
+```
+kubectl create -f <(istioctl kube-inject -f nginx-httpbin.yaml)
+```
+
+To test from outside the cluster we will need the IP address of a Kubernetes node, as well
+as the exposed NodePort port on that node.
+
+```
```
Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: Please use ```shell on all command-line examples. This allows Github and Gogland to do syntax highlighting for the content.

File: docs/tasks/timeout-injection.md [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+curl -i $IP:$NGINX_NODEPORT/status/500
+time curl -i $IP:$NGINX_NODEPORT/delay/5
+```
+
+At this point we have two services.  Both are wired through the Istio service mesh.  Both output the same data.
+
+## Manipulating the service mesh with route rules and destination polices
+
+Istio has many advanced capabilities to control how network traffic flows between microservices.  One of the
+most simple capabilities is to add a timeout.
```
Comment (geeknoid) [inline docs/tasks/timeout-injection.md]: most simple -> simplest

**Pull request merged**